protocols/autonatv2/src/server/behaviour.rs

```diff
@@ -0,0 +1,235 @@
+use std::fmt::Debug;
+use std::{
+    collections::{HashMap, VecDeque},
+    task::{Context, Poll},
+};
+
+use crate::server::handler::dial_request::DialBack;
+use either::Either;
+use libp2p_core::multiaddr::Protocol;
+use libp2p_core::{transport::PortUse, Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::dial_opts::PeerCondition;
+use libp2p_swarm::{
+    dial_opts::DialOpts, ConnectionDenied, ConnectionHandler, ConnectionId, DialError, DialFailure,
+    FromSwarm, NetworkBehaviour, NotifyHandler, ToSwarm,
+};
+use rand_core::{OsRng, RngCore};
+
+use super::handler::{
+    dial_back,
+    dial_request::{self, DialBackCommand},
+    Handler,
+};
+
+pub struct Behaviour<R = OsRng>
+where
+    R: Clone + Send + RngCore + 'static,
+{
+    handlers: HashMap<(Multiaddr, PeerId), ConnectionId>,
+    dialing_dial_back: HashMap<(Multiaddr, PeerId), VecDeque<DialBackCommand>>,
+    pending_events: VecDeque<
+        ToSwarm<
+            <Self as NetworkBehaviour>::ToSwarm,
+            <<Self as NetworkBehaviour>::ConnectionHandler as ConnectionHandler>::FromBehaviour,
+        >,
+    >,
+    rng: R,
+}
+
+impl Default for Behaviour<OsRng> {
+    fn default() -> Self {
+        Self::new(OsRng)
+    }
+}
+
+impl<R> Behaviour<R>
+where
+    R: RngCore + Send + Clone + 'static,
+{
+    pub fn new(rng: R) -> Self {
+        Self {
+            handlers: HashMap::new(),
+            dialing_dial_back: HashMap::new(),
+            pending_events: VecDeque::new(),
+            rng,
+        }
+    }
+
+    fn poll_pending_events(
+        &mut self,
+        _cx: &mut Context<'_>,
+    ) -> Poll<
+        ToSwarm<
+            <Self as NetworkBehaviour>::ToSwarm,
+            <<Self as NetworkBehaviour>::ConnectionHandler as ConnectionHandler>::FromBehaviour,
+        >,
+    > {
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        Poll::Pending
+    }
```

Comment:
With a signature this long, I don't think this function is pulling its weight, I'd suggest inlining it. Especially because you have to use branching again anyway when you call it.

Reply:
Yes, I think that's an artifact too.